extensions/ql-vscode/test/unit-tests/common/invocation-rate-limiter.test.ts

```diff
@@ -0,0 +1,149 @@
+import type { Memento } from "vscode";
+import { InvocationRateLimiter } from "../../../src/common/invocation-rate-limiter";
+
+describe("Invocation rate limiter", () => {
+  // 1 January 2020
+  let currentUnixTime = 1577836800;
+
+  function createDate(dateString?: string): Date {
+    if (dateString) {
+      return new Date(dateString);
+    }
+    const numMillisecondsPerSecond = 1000;
+    return new Date(currentUnixTime * numMillisecondsPerSecond);
+  }
+
+  function createInvocationRateLimiter<T>(
+    funcIdentifier: string,
+    func: () => Promise<T>,
+  ): InvocationRateLimiter<T> {
+    return new InvocationRateLimiter(
+      new MockGlobalStorage(),
+      funcIdentifier,
+      func,
+      (s) => createDate(s),
+    );
+  }
+
+  class MockMemento implements Memento {
+    keys(): readonly string[] {
+      throw new Error("Method not implemented.");
+    }
+    map = new Map<any, any>();
+
+    /**
+     * Return a value.
+     *
+     * @param key A string.
+     * @param defaultValue A value that should be returned when there is no
+     * value (`undefined`) with the given key.
+     * @return The stored value or the defaultValue.
+     */
+    get<T>(key: string, defaultValue?: T): T {
+      return this.map.has(key) ? this.map.get(key) : defaultValue;
+    }
+
+    /**
+     * Store a value. The value must be JSON-stringifyable.
+     *
+     * @param key A string.
+     * @param value A value. MUST not contain cyclic references.
+     */
+    async update(key: string, value: any): Promise<void> {
+      this.map.set(key, value);
+    }
+  }
+
+  class MockGlobalStorage extends MockMemento {
+    public setKeysForSync(_keys: string[]): void {
+      return;
+    }
+  }
```

Comment:
I see that this is from the previous testing implementation, but do we need this? It seems like just having `MockMemento` is enough.

Reply:
Ah, I see why we had this. It's because the real type defined on the `ExtensionContext` is
```typescript
readonly globalState: Memento & {
	/**
	 * Set the keys whose values should be synchronized across devices when synchronizing user-data
	 * like configuration, extensions, and mementos.
	 *
	 * Note that this function defines the whole set of keys whose values are synchronized:
	 *  - calling it with an empty array stops synchronization for this memento
	 *  - calling it with a non-empty array replaces all keys whose values are synchronized
	 *
	 * For any given set of keys this function needs to be called only once but there is no harm in
	 * repeatedly calling it.
	 *
	 * @param keys The set of keys whose values are synced.
	 */
	setKeysForSync(keys: readonly string[]): void;
};
```
So we had to implement `setKeysForSync` too. But now we've declared that `InvocationRateLimiter` only takes a `Memento` instead of `typeof ExtensionContext.globalState`, we don't have to mock this method in the tests anymore.